### PR TITLE
CI Improvements

### PR DESCRIPTION
**Pull request by mrdrogdrog**
This PR improves the build of the docker images by:
- building for amd64, armv7 and armv8
- better caching
- improved multi-step building
- building images for PRs
- cancelling concurrent build processes

### CODE REVIEW

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -45,6 +45,6 @@ jobs:
           file: ./${{ matrix.base }}/Dockerfile
           build-args: |
             VERSION=${{ env.HEDGEDOC_VERSION }}
-          platforms: linux/amd64,linux/arm64,linux/arm/v7
+          platforms: linux/amd64
```
Comment (SISheogorath) [inline .github/workflows/release.yml:48]: Is this change intended?

Comment (mrdrogdrog) [reply .github/workflows/release.yml:48]: Yes. It is explained in the commit message.

Comment (SISheogorath) [reply .github/workflows/release.yml:48]: I see, someone fiddled with production releases… Well…

Comment (mrdrogdrog) [reply .github/workflows/release.yml:48]: Well. It would work if github actions wouldn't time out randomly 🤷

File: .github/workflows/nightly.yml [hunk 2]
```diff
@@ -41,10 +50,12 @@ jobs:
           context: .
           file: ./${{ matrix.base }}/Dockerfile
           labels: quay.expires-after=1w
+          cache-from: type=gha
+          cache-to: type=gha
           build-args: |
             VERSION=${{ env.HEDGEDOC_VERSION }}
-          platforms: linux/amd64
-          push: true
+          platforms: ${{ matrix.arch }}
```
Comment (ErikMichelson) [inline .github/workflows/nightly.yml:57]: Doesn't this build the image three times due to the matrix but instead of pushing a manifest with three sub-images, this will override the latest image with one that only supports linux/arm/v7?

**Review by ErikMichelson — APPROVED** [review]

**Review by SISheogorath — APPROVED** [review]
